docs/proposals/metrics-api.md

```diff
@@ -0,0 +1,134 @@
+
+# Resource Metrics API
+
+*This proposal is based on and supersedes [compute-resource-metrics-api.md](compute-resource-metrics-api.md).*
+
+This document describes API part of MVP version of Resource Metrics API effort in Kubernetes.
+Once the agreement will be made the document will be extended to also cover implementation details.
+This document may be also a subject of changes once we will have more well-defined use cases.
+
+## Goal
+
+The goal for the effort is to provide resource usage metrics for pods and nodes through the API server.
+This will be a stable, versioned API which core Kubernetes components can rely on.
+In the first version only the well-defined use cases should be handled,
+altough the API should be easily extensible for potential future use cases.
+
+## Use cases
+
+This paragraph describes well-defined use cases which should be handled in the first version.
+Use cases which are not listed below are out of the scope of MVP version of Resource Metrics API.
+
+#### Horizontal Pod Autoscaler
+
+HPA needs the latest value of cpu usage (as an mean from some time window)
+for the given set of pods (both pod list or label selector are fine for this purpose).
+
+#### Scheduler
+
+Scheduler in order to schedule best-effort pods requires node level resource usage metrics.
+
+#### kubectl top
+
+This feature is not yet specified/implemented although it seems reasonable to support the following options:
+- ```kubectl top node <id>``` - list all pods running on the given node accompanied with theirs cpu/mem usage
+- ```kubectl top pod <id>``` - print cpu usage stats for the given pod
+- ```kubectl top pod --selector=<label-query>``` - list all pods defined by the label query accompanied with theirs cpu/mem usage
+
+#### Kubernetes dashboard
+
+[Kubernetes dashboard](https://github.com/kubernetes/dashboard) in order to draw graps requires resource usage
+in timeseries format form the last 15 minutes. Alternatively just the latest usage value is fine.
+
+## Proposed API
+
+The metrics API will be in a separate [API group](api-group.md) called ```metrics```.
+
+#### Schema
+
+The proposed schema is as follow. Each object has `TypeMeta` and `ObjectMeta` fields
+to be compatible with Kubernetes API standards.
+
+```go
+type Node struct {
+  TypeMeta
+  ObjectMeta
+  Metrics Metrics
+}
+
+type Pod struct {
+  TypeMeta
+  ObjectMeta
+  Metrics Metrics
+}
+
+// The latest available metrics for the appriopriate resource.
+type Metrics struct {
+  // Start and End specified the time window from which the response was returned.
+  Start unversioned.Time `json:"start"`
+  End   unversioned.Time `json:"end"`
+  Usage ResourceUsage `json:"usage"`
+}
+ 
+type ResourceUsage map[resource.Type]resource.Quantity
+```
+
+By default `Usage` is the mean from the returned time window.
+In the future we may want to introduce the following params:
+`aggreator` (`max`, `min`, `95th`, etc.) and `window` (`1h`, `1d`, `1w`, etc.)
```

Comment:
The resolution of metrics is configurable in Heapster (and possible in any other Metrics provider). The `EndTime` field will let the caller know how fresh the data is while `StartTime` will provide information about the time window. I can't see how `StartTime/EndTime` and `window` are incompatible (see examples below). Also probably `EndTime` in response when querying with `window` parameter specified although I thinks the response should be self-descriptive.

Few examples:
 1. Heapster resolution is 1m, everything works correctly. At 10:01:10 the response might be: `{StartTime: 10:00:00, EndTime: 10:01:00}`.
 2. Heapster resolution is 1m, the is some lag while collecting metrics in Heapster. At 10:01:10 the response might be: `{StartTime: 9:58:00, EndTime: 9:59:00}`.
 3. Heapster resolution is 1m, Heapster uses some persistent storage for historical data. Querying with window specified to 1h. At 10:01:10 the response might be: `{StartTime: 9:01:00, EndTime: 10:01:00}`.
 4. Heapster resolution is 1m, Heapster remembers only data from last 15 minutes in memory. Querying with window specified to 1h. At 10:01:10 the response might be: `{StartTime: 9:01:00, EndTime: 10:01:00}`. Alternatively the error might be returned due to lack of data.